python/tvm/relay/backend/contrib/tidl.py

```diff
@@ -0,0 +1,1403 @@
+                outputs.append(out)
+        else:
+            outputs.append(visit_body)
+        # Create new function with added subgraph inputs + outputs
+        return relay.Function(expr.params, relay.Tuple(outputs + self.additional_outputs))
+
+class RemoveMultiplyByOne(ExprMutator):
+    """
+    Removes multiply by 1.0f. This pass when followed by
+    RemoveRedundantTranspose is intended to remove a pattern of
+    Transpose([1, 0]) -> Scale(1.0f) -> Transpose([1, 0]) produced by
+    PyTorch's addmm operator.
+    """
+    def visit_call(self, expr):
+        if expr.op.name == "multiply":
+            if isinstance(expr.args[1], tvm.relay.expr.Constant):
+                data = expr.args[1].data.asnumpy()
+                if data.shape == () and data.item() == 1.0:
+                    return expr.args[0]
+            if isinstance(expr.args[0], tvm.relay.expr.Constant):
+                data = expr.args[0].data.asnumpy()
+                if data.shape == () and data.item() == 1.0:
+                    return expr.args[1]
+        return super().visit_call(expr)
+
+def generate_subgraph_tensors(tidl_target, mod, params, input_node, input_data):
+    """
+    """
+
+    # From partitioned module, create a "calibration model" which can be
+    # executed on CPU and will give additional outputs for boundary tensors.
+    mod_tvm = relay.transform.InferType()(mod)
+    mod_tvm = relay.transform.Inline()(mod_tvm)
+    my_mutator = CalibrationGraphMutator(tidl_target)
+    mod_tvm["main"] = my_mutator.make_calibration_graph(mod_tvm["main"])
+    #print("Calibration module:", mod_tvm)
+    print("Input map:", my_mutator.name_map)
+
+    # Build and execute calibration graph to get outputs
+    # Use opt_level=0 to avoid optimizations which modify the module (could change original module)
+    with relay.build_config(opt_level=0):
+        graph, lib, params = relay.build(mod_tvm, "llvm", params=params)
+    mod = graph_runtime.create(graph, lib, ctx=tvm.cpu(0))
+    mod.set_input(input_node, input_data)
+    mod.set_input(**params)
+    mod.run()
+    #mod.run(data=input_data, weight1=params_w1, weight2=params_w2)
+
+    results = [mod.get_output(i).asnumpy() for i in range(mod.get_num_outputs())]
+    np.savetxt('graph_output.txt', results[0].flatten(), fmt='%10.5f')
+
+    # We now have subgraph inputs
+    # {1: 'tidl_1_i0', 2: 'tidl_1_o0', 3: 'tidl_0_i0', 4: 'tidl_0_o0'}
+    subgraph_tensors = {}
+    for i in range(len(results)):
+        if i in my_mutator.name_map:
+            subgraph_tensors[my_mutator.name_map[i]]=results[i]
+            #print("Subgraph input: ", my_mutator.name_map[i], " tensor: ", results[i])
+            file_name = my_mutator.name_map[i] + ".txt"
+            np.savetxt(file_name, results[i].flatten(), fmt='%10.5f')
+
+    for key, value in subgraph_tensors.items():
+        print("Subgraph tensor: ", key, value.shape)
+
+    return subgraph_tensors
+
+class VarReplacer(ExprMutator):

```

Comment:
I think this is a duplicate, already defined on line 1051.

Reply:
Removed.